pkg/controllers/disruption/tracker.go

```diff
@@ -0,0 +1,276 @@
+	bucketsThresholdMap map[string]*BucketThresholds
+	enabled             bool
+
+	client      client.Client
+	cluster     *state.Cluster
+	clusterCost *cost.ClusterCost
+	clock       clock.Clock
+}
+
+// NewTracker creates a queue that will asynchronously track and emit metrics on decisions
+func NewTracker(cluster *state.Cluster, clock clock.Clock, bucketsThresholdMap map[string]*BucketThresholds, enabled bool) *Tracker {
+	decisionCache := cache.New(defaultExpiration, cleanupInterval)
+	decisionCache.OnEvicted(DecisionExpired)
+	for _, st := range lo.Values(bucketsThresholdMap) {
+		sort.Slice(st.thresholds, func(i, j int) bool {
+			return st.thresholds[i].lowerBound < st.thresholds[j].lowerBound
+		})
+	}
+	for _, b := range DecisionDimensions {
+		_, exists := bucketsThresholdMap[b]
+		if !exists {
+			log.FromContext(context.Background()).Error(fmt.Errorf("missing thresholds for necessary bucket"), "bucket", b)
+			DecisionTrackerErrors.Inc(map[string]string{})
+			// disabling to prevent panics later.
+			enabled = false
+		}
+	}
+	tracker := &Tracker{
+		// nolint:staticcheck
+		decisionCache:       decisionCache,
+		disruptionKeyMap:    make(map[string]string),
+		bucketsThresholdMap: bucketsThresholdMap,
+		enabled:             enabled,
+		cluster:             cluster,
+		clock:               clock,
+	}
+	return tracker
+}
+
+func (t *Tracker) GatherClusterState(ctx context.Context) (*ClusterStateSnapshot, error) {
+	deployments := appsv1.DeploymentList{}
+	err := t.client.List(ctx, &deployments)
+	if err != nil {
+		return nil, err
+	}
+
+	totalDesiredPodCount := 0
+	totalDesiredPodCount = lo.SumBy(deployments.Items, func(d appsv1.Deployment) int {
+		return int(*d.Spec.Replicas)
+	})
+
+	rs := lo.Map(deployments.Items, func(d appsv1.Deployment, _ int) corev1.ResourceList {
+		requests := lo.Map(d.Spec.Template.Spec.Containers, func(c corev1.Container, _ int) corev1.ResourceList {
+			return c.Resources.Requests
+		})
+
+		totalContainerResources := resources.Merge(requests...)
+
+		for i, r := range totalContainerResources {
+			r.Set(r.Value() * int64(*d.Spec.Replicas))
+			totalContainerResources[i] = r
+		}
+
+		return totalContainerResources
+	})
+
+	totalDesiredPodResources := resources.Merge(rs...)
+
+	return &ClusterStateSnapshot{
+		clusterCost:              t.clusterCost.GetClusterCost(),
+		podResources:             t.cluster.GetTotalPodResourceRequests(),
+		totalNodes:               len(t.cluster.DeepCopyNodes()),
+		totalDesiredPodCount:     totalDesiredPodCount,
+		totalDesiredPodResources: &totalDesiredPodResources,
+	}, nil
+}
+
+func (t *Tracker) AddCommand(ctx context.Context, cmd *Command) {
+	if !t.enabled {
+		return
+	}
+	s, err := t.GatherClusterState(ctx)
+	if err != nil {
+		log.FromContext(ctx).Error(err, "gathering cluster state", "cmd", cmd)
+		DecisionTrackerErrors.Inc(map[string]string{})
+	}
+	d := TrackedDecision{
+		clusterState: s,
+		startTime:    t.clock.Now(),
+		command:      cmd,
```

Comment:
I wonder what the memory impact of storing the entire command in the TrackedDecision will be. The main concern would be that we're storing a set of deep copied `StateNodes` inside the command. Do we need to store the whole command?